src/darwin/Framework/CHIP/MTRLogging.mm

```diff
@@ -0,0 +1,184 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <MTRLogging.h>
+#import <lib/support/logging/CHIPLogging.h>
+#import <os/log.h>
+#import <pthread.h>
+
+// MARK: - OsLogMTRLogger
+
+/**
+ * An implementation of the MTRLogger protocol that logs messages to `os_log`.
+ */
+@interface OsLogMTRLogger : NSObject <MTRLogger>
+
+// Disable construction
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+
+/**
+ * @returns The singleton OsLogMTRLogger
+ */
++ (instancetype)sharedInstance;
+
+@end
+
+@implementation OsLogMTRLogger
+
++ (instancetype)sharedInstance
+{
+    static OsLogMTRLogger * singleton = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        // initialize the factory.
+        singleton = [[OsLogMTRLogger alloc] init];
+    });
+    return singleton;
+}
+
+- (instancetype)init
+{
+    return self;
+}
+
+- (BOOL)isLoggingEnabledAtLevel:(MTRLogLevel)level
+{
+    return true;
+}
+
+- (void)logAtLevel:(MTRLogLevel)level message:(NSString *)message
+{
+    switch (level) {
+    case MTRLogLevelFault:
+        os_log_fault(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelError:
+        os_log_error(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelInfo:
+        os_log_info(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelDebug:
+        os_log_debug(OS_LOG_DEFAULT, "%@", message);
+        break;
+    default:
+        break;
+    }
+}
+
+@end
+
+// MARK: - MTRLogging
+
+@implementation MTRLoggingConfiguration
+
+static id<MTRLogger> singleton_logger;
+
+/**
+ * A log redirection function that will be used to capture logs from the Matter C++ SDK.
+ */
+static ENFORCE_FORMAT(3, 0) void cpp_log_redirect_callback(const char * module, uint8_t category, const char * msg, va_list args)
+{
+    // Pre-format the message. Additional information will be prepended and appended before
+    // logging.
+    char formatted_msg[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE];
+    int32_t formatted_msg_length = vsnprintf(formatted_msg, sizeof(formatted_msg), msg, args);
+    if (formatted_msg_length < 0) {
+        // In the event that we could not format the message, simply copy over the
+        // unformatted message. It won't contain all the information that we need, but
+        // it's better than dropping a log on the floor.
+        strncpy(formatted_msg, msg, sizeof(formatted_msg));
+        return;
+    }
+
+    // Explicitly NULL-terminate the formatted string for safety.
+    formatted_msg[sizeof(formatted_msg) - 1] = '\0';
+
+    // Collect the process and thread information to match the original CHIPLogging implementation
+    // in src/platform/Darwin/Logging.cpp
```

Comment:
Can this share code with that implementation, ideally, instead of duplicating it and needing the two to stay in sync but having no indication over there that this code even exists?

Reply:
Yes, moved some common code together